src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineControl.xaml.cs

```diff
@@ -0,0 +1,291 @@
+using Microsoft.CodeAnalysis.Editor.Tagging;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.VisualStudio.Editor;
+using Microsoft.VisualStudio.LanguageServices.Implementation;
+using Microsoft.VisualStudio.LanguageServices.Implementation.LanguageServiceBrokerShim;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.TextManager.Interop;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.DocumentOutline
+{
+    /// <summary>
+    /// Interaction logic for DocumentOutlineControl.xaml
+    /// </summary>
+    internal partial class DocumentOutlineControl : UserControl, IVsCodeWindowEvents, IDisposable
+    {
+        private readonly ILanguageServiceBrokerShim _languageServiceBroker;
+        private readonly IThreadingContext _threadingContext;
+        private readonly IAsynchronousOperationListener _asyncListener;
+        private readonly IVsEditorAdaptersFactoryService _editorAdaptersFactoryService;
+        private readonly IVsCodeWindow _codeWindow;
+        private readonly ComEventSink _codeWindowEventsSink;
+        private readonly CompilationAvailableTaggerEventSource _textViewEventSource;
+        private readonly CancellationTokenSource _cancellationTokenSource;
+        private readonly CancellationToken _cancellationToken;
+
+        /// <summary>
+        /// The type of sorting to be applied to the data model in <see cref="FilterAndSortDataModelAsync"/>.
+        /// </summary>
+        /// <remarks>
+        /// It is only safe to read/mutate SortOption from the UI thread.
+        /// </remarks>
+        private SortOption SortOption
+        {
+            get
+            {
+                _threadingContext.ThrowIfNotOnUIThread();
+                return _sortOption;
+            }
+            set
+            {
+                _threadingContext.ThrowIfNotOnUIThread();
+                _sortOption = value;
+            }
+        }
+
+        private SortOption _sortOption;
+
+        /// <summary>
+        /// Queue to batch up work to do to compute the data model. Used so we can batch up a lot of events 
+        /// and only fetch the model once for every batch. The bool parameter is unused.
+        /// </summary>
+        private readonly AsyncBatchingWorkQueue<bool, DocumentSymbolDataModel?> _computeDataModelQueue;
+
+        /// <summary>
+        /// Queue to batch up work to do to filter and sort the data model. The bool parameter is unused.
+        /// </summary>
+        private readonly AsyncBatchingWorkQueue<bool, DocumentSymbolDataModel?> _filterAndSortDataModelQueue;
+
+        /// <summary>
+        /// Queue to batch up work to do to highlight the currently selected symbol node, expand/collapse nodes,
+        /// then update the UI.
+        /// </summary>
+        private readonly AsyncBatchingWorkQueue<ExpansionOption> _highlightExpandAndPresentItemsQueue;
```

Comment:
nit: it would be good to have a comment somewhere stating the relationship between the queues.  e.g. we start with the computeData queue which will then flow into the filtersortqueue which will then flowinto the highlightpresent queue.  you can also state what user operations flow into which queues.